File: backend/app/services/task_service.py
"""Task service - implements business logic for the task CRUD + workflow API."""
from __future__ import annotations

import logging
from typing import List, Optional

from fastapi import HTTPException, status as http_status
from sqlalchemy.orm import Session

from app.models.audit_log import AuditAction
from app.models.task import Task, TaskPriority, TaskStatus
from app.models.user import User
from app.repositories.task_repository import TaskRepository
from app.repositories.user_repository import UserRepository
from app.schemas.task import TaskCreate, TaskUpdate
from app.services.approval_service import ApprovalService
from app.services.audit_log_service import AuditLogService
from app.services.workflow_service import WorkflowService


logger = logging.getLogger(__name__)


class TaskService:
    """Encapsulates CRUD and lifecycle business rules for tasks."""

    def __init__(self, db: Session) -> None:
        self.db = db
        self.repo = TaskRepository(db)
        self.users = UserRepository(db)
        self.workflow = WorkflowService(db)
        self.approvals = ApprovalService(db)
        self.audit = AuditLogService(db)

    def list_tasks(
        self,
        status: Optional[TaskStatus] = None,
        priority: Optional[TaskPriority] = None,
        assigned_to_user_id: Optional[int] = None,
    ) -> List[Task]:
        return self.repo.list_tasks(
            status=status,
            priority=priority,
            assigned_to_user_id=assigned_to_user_id,
        )

    def get_task_or_404(self, task_id: int) -> Task:
        task = self.repo.get(task_id)
        if task is None:
            raise HTTPException(
                status_code=http_status.HTTP_404_NOT_FOUND,
                detail=f"Task {task_id} not found.",
            )
        return task

    def _ensure_creator_or_manager(self, task: Task, user: User, *, action: str) -> None:
        """Analysts may only act on tasks they created; managers may act on any."""
        if user.is_analyst and task.created_by_user_id != user.id:
            raise HTTPException(
                status_code=http_status.HTTP_403_FORBIDDEN,
                detail=f"Analysts can only {action} tasks they created.",
            )

    def create_task(self, data: TaskCreate, current_user: User) -> Task:
        assignee = self.users.get(data.assigned_to_user_id)
        if assignee is None:
            raise HTTPException(
                status_code=http_status.HTTP_400_BAD_REQUEST,
                detail=f"Assignee user {data.assigned_to_user_id} not found.",
            )
        task = Task(
            title=data.title.strip(),
            description=data.description or "",
            priority=data.priority,
            status=TaskStatus.PENDING,
            assigned_to_user_id=assignee.id,
            created_by_user_id=current_user.id,
        )
        self.repo.add(task)
        self.workflow.record_initial_event(task=task, changed_by=current_user)
        self.audit.log_action(
            user_id=current_user.id,
            action=AuditAction.CREATE_TASK,
            entity_type="task",
            entity_id=task.id,
            details=f"Created task '{task.title}' assigned to user {assignee.id}",
        )
        self.db.commit()
        self.db.refresh(task)
        return task

    def update_task(self, task_id: int, data: TaskUpdate, current_user: User) -> Task:
        task = self.get_task_or_404(task_id)
        if task.status == TaskStatus.COMPLETED:
            raise HTTPException(
                status_code=http_status.HTTP_400_BAD_REQUEST,
                detail="Completed tasks cannot be edited.",
            )
        self._ensure_creator_or_manager(task, current_user, action="edit")

        changed_fields: List[str] = []
        if data.title is not None and data.title.strip() != task.title:
            task.title = data.title.strip()
            changed_fields.append("title")
        if data.description is not None and data.description != task.description:
            task.description = data.description
            changed_fields.append("description")
        if data.priority is not None and data.priority != task.priority:
            task.priority = data.priority
            changed_fields.append("priority")
        if (
            data.assigned_to_user_id is not None
            and data.assigned_to_user_id != task.assigned_to_user_id
        ):
            assignee = self.users.get(data.assigned_to_user_id)
            if assignee is None:
                raise HTTPException(
                    status_code=http_status.HTTP_400_BAD_REQUEST,
                    detail=f"Assignee user {data.assigned_to_user_id} not found.",
                )
            task.assigned_to_user_id = assignee.id
            changed_fields.append("assigned_to_user_id")

        self.audit.log_action(
            user_id=current_user.id,
            action=AuditAction.UPDATE_TASK,
            entity_type="task",
            entity_id=task.id,
            details=(
                f"Updated fields: {', '.join(changed_fields)}"
                if changed_fields
                else "No-op update"
            ),
        )
        self.db.commit()
        self.db.refresh(task)
        return task

    def delete_task(self, task_id: int, current_user: User) -> None:
        task = self.get_task_or_404(task_id)
        if task.status in (TaskStatus.APPROVED, TaskStatus.COMPLETED):
            raise HTTPException(
                status_code=http_status.HTTP_400_BAD_REQUEST,
                detail="Approved or completed tasks cannot be deleted.",
            )
        self._ensure_creator_or_manager(task, current_user, action="delete")
        task_id_value = task.id
        title = task.title
        self.repo.delete(task)
        self.audit.log_action(
            user_id=current_user.id,
            action=AuditAction.DELETE_TASK,
            entity_type="task",
            entity_id=task_id_value,
            details=f"Deleted task '{title}'",
        )
        self.db.commit()

    def submit_task(self, task_id: int, current_user: User, comment: Optional[str] = None) -> Task:
        task = self.get_task_or_404(task_id)
        self._ensure_creator_or_manager(task, current_user, action="submit")
        self.workflow.transition(
            task=task,
            to_status=TaskStatus.SUBMITTED,
            changed_by=current_user,
            note=comment or "Submitted for review",
        )
        self.audit.log_action(
            user_id=current_user.id,
            action=AuditAction.SUBMIT_TASK,
            entity_type="task",
            entity_id=task.id,
            details=comment or "Submitted task",
        )
        self.db.commit()
        self.db.refresh(task)
        return task

    def approve_task(self, task_id: int, manager: User, comment: Optional[str] = None) -> Task:
        task = self.get_task_or_404(task_id)
        self.approvals.approve_task(task=task, manager=manager, comment=comment)
        self.audit.log_action(
            user_id=manager.id,
            action=AuditAction.APPROVE_TASK,
            entity_type="task",
            entity_id=task.id,
            details=comment or "Approved task",
        )
        self.db.commit()
        self.db.refresh(task)
        return task

    def reject_task(self, task_id: int, manager: User, comment: Optional[str] = None) -> Task:
        task = self.get_task_or_404(task_id)
        self.approvals.reject_task(task=task, manager=manager, comment=comment)
        self.audit.log_action(
            user_id=manager.id,
            action=AuditAction.REJECT_TASK,
            entity_type="task",
            entity_id=task.id,
            details=comment or "Rejected task",
        )
        self.db.commit()
        self.db.refresh(task)
        return task

    def complete_task(self, task_id: int, manager: User, comment: Optional[str] = None) -> Task:
        task = self.get_task_or_404(task_id)
        if not manager.is_manager:
            raise HTTPException(
                status_code=http_status.HTTP_403_FORBIDDEN,
                detail="Only managers can complete tasks.",
            )
        self.workflow.transition(
            task=task,
            to_status=TaskStatus.COMPLETED,
            changed_by=manager,
            note=comment or "Completed by manager",
        )
        self.audit.log_action(
            user_id=manager.id,
            action=AuditAction.COMPLETE_TASK,
            entity_type="task",
            entity_id=task.id,
            details=comment or "Completed task",
        )
        self.db.commit()
        self.db.refresh(task)
        return task

File: backend/src/main/java/com/example/bwms/service/TaskService.java
// Translated from: backend/app/services/task_service.py
package com.example.bwms.service;

import com.example.bwms.dto.TaskCreateDto;
import com.example.bwms.dto.TaskUpdateDto;
import com.example.bwms.model.*;
import com.example.bwms.repository.TaskRepository;
import com.example.bwms.repository.UserRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.criteria.JoinType;
import jakarta.persistence.criteria.Predicate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.http.HttpStatus;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.ArrayList;
import java.util.List;

@Service
public class TaskService {

    private static final Logger log = LoggerFactory.getLogger(TaskService.class);

    @PersistenceContext
    private EntityManager entityManager;

    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final WorkflowService workflowService;
    private final ApprovalService approvalService;
    private final AuditLogService auditLogService;

    public TaskService(TaskRepository taskRepository, UserRepository userRepository,
                       WorkflowService workflowService, ApprovalService approvalService,
                       AuditLogService auditLogService) {
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.workflowService = workflowService;
        this.approvalService = approvalService;
        this.auditLogService = auditLogService;
    }

    @Transactional(readOnly = true)
    public List<Task> listTasks(@Nullable TaskStatus status,
                                @Nullable TaskPriority priority,
                                @Nullable Long assignedToUserId) {
        Specification<Task> spec = (root, query, cb) -> {
            // Skip eager fetches on the count query Spring may issue
            if (query.getResultType() == Task.class) {
                root.fetch("assignee", JoinType.LEFT);
                root.fetch("creator", JoinType.LEFT);
            }
            List<Predicate> predicates = new ArrayList<>();
            if (status != null)             predicates.add(cb.equal(root.get("status"), status));
            if (priority != null)           predicates.add(cb.equal(root.get("priority"), priority));
            if (assignedToUserId != null)   predicates.add(cb.equal(root.get("assignedToUserId"), assignedToUserId));
            return predicates.isEmpty() ? cb.conjunction() : cb.and(predicates.toArray(new Predicate[0]));
        };
        return taskRepository.findAll(spec, Sort.by(Sort.Direction.DESC, "createdAt"));
    }

    @Transactional(readOnly = true)
    public Task getTaskOr404(Long taskId) {
        return taskRepository.findById(taskId).orElseThrow(() ->
                new ResponseStatusException(HttpStatus.NOT_FOUND,
                        "Task " + taskId + " not found."));
    }

    // Python: TaskService._ensure_creator_or_manager
    private void ensureCreatorOrManager(Task task, User user, String action) {
        if (user.isAnalyst() && !task.getCreatedByUserId().equals(user.getId())) {
            throw new ResponseStatusException(
                    HttpStatus.FORBIDDEN,
                    "Analysts can only " + action + " tasks they created.");
        }
    }

    @Transactional
    public Task createTask(TaskCreateDto data, User currentUser) {
        User assignee = userRepository.findById(data.assignedToUserId()).orElseThrow(() ->
                new ResponseStatusException(HttpStatus.BAD_REQUEST,
                        "Assignee user " + data.assignedToUserId() + " not found."));

        Task task = new Task();
        task.setTitle(data.title().strip());
        task.setDescription(data.description() != null ? data.description() : "");
        task.setPriority(data.priority() != null ? data.priority() : TaskPriority.MEDIUM);
        task.setStatus(TaskStatus.PENDING);
        task.setAssignedToUserId(assignee.getId());
        task.setCreatedByUserId(currentUser.getId());
        taskRepository.save(task);

        workflowService.recordInitialEvent(task, currentUser, null);
        auditLogService.logAction(currentUser.getId(), AuditAction.CREATE_TASK,
                "task", task.getId(),
                "Created task '" + task.getTitle() + "' assigned to user " + assignee.getId());

        // Detach from first-level cache so the reload below re-queries with the EntityGraph
        // (without detach, Hibernate returns the cached entity which has null lazy associations)
        entityManager.detach(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    @Transactional
    public Task updateTask(Long taskId, TaskUpdateDto data, User currentUser) {
        Task task = getTaskOr404(taskId);
        if (task.getStatus() == TaskStatus.COMPLETED) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST, "Completed tasks cannot be edited.");
        }
        ensureCreatorOrManager(task, currentUser, "edit");

        List<String> changedFields = new ArrayList<>();
        if (data.title() != null && !data.title().strip().equals(task.getTitle())) {
            task.setTitle(data.title().strip());
            changedFields.add("title");
        }
        if (data.description() != null && !data.description().equals(task.getDescription())) {
            task.setDescription(data.description());
            changedFields.add("description");
        }
        if (data.priority() != null && data.priority() != task.getPriority()) {
            task.setPriority(data.priority());
            changedFields.add("priority");
        }
        if (data.assignedToUserId() != null
                && !data.assignedToUserId().equals(task.getAssignedToUserId())) {
            userRepository.findById(data.assignedToUserId()).orElseThrow(() ->
                    new ResponseStatusException(HttpStatus.BAD_REQUEST,
                            "Assignee user " + data.assignedToUserId() + " not found."));
            task.setAssignedToUserId(data.assignedToUserId());
            changedFields.add("assigned_to_user_id");
        }

        String details = changedFields.isEmpty()
                ? "No-op update"
                : "Updated fields: " + String.join(", ", changedFields);
        auditLogService.logAction(currentUser.getId(), AuditAction.UPDATE_TASK,
                "task", task.getId(), details);

        taskRepository.save(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    @Transactional
    public void deleteTask(Long taskId, User currentUser) {
        Task task = getTaskOr404(taskId);
        if (task.getStatus() == TaskStatus.APPROVED || task.getStatus() == TaskStatus.COMPLETED) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST, "Approved or completed tasks cannot be deleted.");
        }
        ensureCreatorOrManager(task, currentUser, "delete");

        Long taskIdValue = task.getId();
        String title = task.getTitle();
        taskRepository.delete(task);
        auditLogService.logAction(currentUser.getId(), AuditAction.DELETE_TASK,
                "task", taskIdValue, "Deleted task '" + title + "'");
    }

    @Transactional
    public Task submitTask(Long taskId, User currentUser, @Nullable String comment) {
        Task task = getTaskOr404(taskId);
        ensureCreatorOrManager(task, currentUser, "submit");
        workflowService.transition(task, TaskStatus.SUBMITTED, currentUser,
                comment != null ? comment : "Submitted for review");
        auditLogService.logAction(currentUser.getId(), AuditAction.SUBMIT_TASK,
                "task", task.getId(), comment != null ? comment : "Submitted task");
        taskRepository.save(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    @Transactional
    public Task approveTask(Long taskId, User manager, @Nullable String comment) {
        Task task = getTaskOr404(taskId);
        approvalService.approveTask(task, manager, comment);
        auditLogService.logAction(manager.getId(), AuditAction.APPROVE_TASK,
                "task", task.getId(), comment != null ? comment : "Approved task");
        taskRepository.save(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    @Transactional
    public Task rejectTask(Long taskId, User manager, @Nullable String comment) {
        Task task = getTaskOr404(taskId);
        approvalService.rejectTask(task, manager, comment);
        auditLogService.logAction(manager.getId(), AuditAction.REJECT_TASK,
                "task", task.getId(), comment != null ? comment : "Rejected task");
        taskRepository.save(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    @Transactional
    public Task completeTask(Long taskId, User manager, @Nullable String comment) {
        Task task = getTaskOr404(taskId);
        if (!manager.isManager()) {
            throw new ResponseStatusException(
                    HttpStatus.FORBIDDEN, "Only managers can complete tasks.");
        }
        workflowService.transition(task, TaskStatus.COMPLETED, manager,
                comment != null ? comment : "Completed by manager");
        auditLogService.logAction(manager.getId(), AuditAction.COMPLETE_TASK,
                "task", task.getId(), comment != null ? comment : "Completed task");
        taskRepository.save(task);
        return taskRepository.findById(task.getId()).orElseThrow();
    }

    public List<WorkflowEvent> getTaskHistory(Long taskId) {
        return workflowService.getTaskHistory(taskId);
    }
}
